### PR TITLE
boards/arduino-mega2560: add TTY_BOARD_FILTER

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This allows automatically selecting TTY actually belonging to an
Arduino Mega2560 if `MOST_RECENT_PORT=1` is set.

### Testing procedure

```
make BOARD=arduino-mega2560 MOST_RECENT_PORT=1 -C examples/default flash term
```

Should automatically select the TTY of the most recently connected Arduino Mega2560, even if a board of a different type is connect more recently.

### Issues/PRs references

- [x] depends on (and includes) https://github.com/RIOT-OS/RIOT/pull/19011

### CODE REVIEW

**Comment by maribu** [comment]
(This should be super quick in Murdock, as `can_fast_ci_run.py` should detect that only one board was changed. Hence, no harm in already having this in the CI.)

**Comment by gschorcht** [comment]
The test fails for me. First I thought that I had to install `pyudev`. But also with `pyudev`, I get the following error:
```
avrdude -c stk500v2 -p m2560 -P  -D -U flash:w:examples/default/bin/arduino-mega2560/default.h

avrdude: ser_open(): can't open device "-D": No such file or directory
```
It seems that the port isn't defined.

**Comment by gschorcht** [comment]
Ah ok, maybe I have to try it on top of PR #19011

**Comment by gschorcht** [comment]
> Ah ok, maybe I have to try it on top of PR #19011

No, same error.

**Comment by maribu** [comment]
Hmm, what is the output of `make list-ttys`?

**Comment by gschorcht** [comment]
> Hmm, what is the output of make list-ttys?

<details>

```
path         | driver | vendor       | model                                | model_db                                             | serial | ctime    | iface_num
-------------|--------|--------------|--------------------------------------|------------------------------------------------------|--------|----------|----------
/dev/ttyUSB1 | ch341  | 1a86         | USB2.0-Serial                        | HL-340 USB-Serial adapter                            | None   | 16:58:16 | 0        
/dev/ttyUSB0 | cp210x | Silicon Labs | CP2102 USB to UART Bridge Controller | CP2102/CP2109 UART Bridge Controller [CP210x family] | 0001   | 16:57:27 | 0        
```

</details>
Obviously my Arduino MEGA 2560 board is a Chineese clone that doesn't use an ATMEGA16U2 for the USB interface but an CH341 USB-Serial adapter :worried: No wonder that the board isn't detected. I had anywhere an original Arduino MEGA 2560  ADK, but I don't know where.

**Comment by maribu** [comment]
I updated the PR a bit to also allow adding custom commands to detect the TTYs instead of just a `TTY_BOARD_FILTER` and extended the `ttys.py` to return an error code if no TTY was found. This allows chaining the `ttys.py` command in shell e.g. like

```sh
ttys.py --filter-genuine || ttys.py --filter-clones
```

And for the Arduino Mega 2560 I now overwrite the shell command to fall back to detecting cheap USB UART bridges when no genuine Arduino Mega 2560 board was found. I think this is the best compromise, as it still reliably detects a genuine Arduino Mega 2560 (which we can tell apart by the UDEV data from other USB UART bridges), but won't break usage for clone boards.

File: doc/doxygen/src/flashing.md [hunk 1]
```diff
@@ -430,4 +430,68 @@ vendor or model to only considered TTYs that actually may belong to the selected
...
+allows to always reliably identify them correctly (e.g. the firmware on the
+ATmega16U2 used as USB to UART converted on Arduino Mega2560 will provide
+identification data unique to that board).
+
+Adding Board Filters
+--------------------
+
+After connecting as many variants of the board you target (and maybe some others
+to test that the filter actually filters out non-matching boards). Then first
+run `./dist/tools/user-serial/ttys.py` without arguments and study the output.
```
Comment (gschorcht) [inline doc/doxygen/src/flashing.md:443]:
```suggestion
run `./dist/tools/usb-serial/ttys.py` without arguments and study the output.
```

**Review by gschorcht — CHANGES_REQUESTED** [review]

**Comment by gschorcht** [comment]
> And for the Arduino Mega 2560 I now overwrite the shell command to fall back to detecting cheap USB UART bridges when no genuine Arduino Mega 2560 board was found. I think this is the best compromise, as it still reliably detects a genuine Arduino Mega 2560 (which we can tell apart by the UDEV data from other USB UART bridges), but won't break usage for clone boards.

But of course this only works if the Arduino Mega 2560 Clone was connected last.

File: doc/doxygen/src/flashing.md [hunk 2]
```diff
@@ -430,4 +430,68 @@ vendor or model to only considered TTYs that actually may belong to the selected
...
+/dev/ttyACM3 | cdc_acm | Arm                      | BBC micro:bit CMSIS-DAP              | ARM mbed                                             | 99053602000528334c41b84da1f2f09d000000006e052820 | 12:21:03 | 1
+/dev/ttyUSB0 | cp210x  | Silicon Labs             | CP2102 USB to UART Bridge Controller | CP2102/CP2109 UART Bridge Controller [CP210x family] | 0001                                             | 16:57:27 | 0
+
+Now we add arguments to the invocation of `ttys.py` to filter the list e.g.
+by model, vendor etc. (note: as regex!) ideally until only the target boards
+are listed. Some boards do not provide enough information to e.g. tell them
+apart from other boards using the same USB to UART bridge or the same debugger.
+In that case we have to live with some "bycatch".
+
+In the case of the Arduino Mega 2560 the paramters
```
Comment (gschorcht) [inline doc/doxygen/src/flashing.md:462]:
```suggestion
In the case of the Arduino Mega 2560 the parameters
```

**Review by gschorcht — CHANGES_REQUESTED** [review]

File: doc/doxygen/src/flashing.md (hunk 1 above)
Comment (gschorcht) [reply doc/doxygen/src/flashing.md:443]: Feel free to squash the fixes directly.

**Review by gschorcht — APPROVED** [review]

**Comment by gschorcht** [comment]
bors merge

**Comment by maribu** [comment]
Squashed.

bors retry

**Comment by maribu** [comment]
bors retry

**Comment by maribu** [comment]
Thx :)

**Comment by gschorcht** [comment]
Oops, I accidentally restarted a full build in the wrong browser window for this PR.

**Pull request merged**